### ISSUE TITLE
Run fullscreen operations in parallel to script event-loop

### ISSUE THREAD

**Issue by gterzian** [servo/servo#21600]
It appears to me that the enter- and exit-fullscreen operation could use some restructuring.

Currently, these are done via a task queued on the main event-loop, see for example https://github.com/servo/servo/blob/75c0bdaf1b3a26ee8ae50d4cf3fb6b7546d3ccb8/components/script/dom/document.rs#L2873, which then, in some conditional branches at least, results in the script thread sending a message to and waiting on the reply from the layout thread [in Window.force_reflow](https://github.com/servo/servo/blob/75c0bdaf1b3a26ee8ae50d4cf3fb6b7546d3ccb8/components/script/dom/window.rs#L1372)

The task is then effectively blocked  waiting on layout, and when the reply has been received, the promise will be resolved at the end of the task, see https://github.com/servo/servo/blob/75c0bdaf1b3a26ee8ae50d4cf3fb6b7546d3ccb8/components/script/dom/element.rs#L3218

The [spec](https://fullscreen.spec.whatwg.org/#exit-fullscreen)(see Step : "Return promise, and run the remaining steps in parallel.") mentions that those steps should be done "in parallel". 

Therefore, the "correct" design would appear to be sending a message to the layout thread, returning the promise immediately, and when the parallel layout steps are done, queueing a task back on the script event-loop and  resolving the promise. 

A TODO has been added at https://github.com/servo/servo/pull/21583/files#diff-30a18e04d7e0b66aafdf192e416cad44R2840

**Comment by gterzian** [comment]
By the way, the HTML spec does specific how to integrate steps from spec other than the HTML one(such as the fullscreen one) with the HTML event-loop, at https://html.spec.whatwg.org/multipage/#event-loop-for-spec-authors

**Comment by pshaughn** [comment]
WPT tests expect fullscreen state to be something they can mess around with as part of iframe sandbox testing, in html/semantics/embedded-content/the-iframe-element/iframe-allowfullscreen and iframe-allow. These test failures could be as much about fullscreen as about iframes.

**Comment by jdm** [comment]
Now that the layout thread is no longer a separate thread, I don't know if there's actually much to do here.

**Comment by mrobinson** [comment]
I guess we just need to move the fullscreen steps to the appropriate place in "update the rendering" here: https://github.com/servo/servo/blob/1b4c281e103c7d4f0c2de6e0f2df4f33fcc43b2f/components/script/script_thread.rs#L1181

**Comment by mrobinson** [comment]
I think this is effectively a duplicate of #42067.
